addons/hr_expense/models/ir_attachment.py

```diff
@@ -0,0 +1,39 @@
+from odoo import models, api
+
+class IrAttachment(models.Model):
+    _inherit = 'ir.attachment'
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        attachments = super().create(vals_list)
+        if self.env.context.get('sync_attachment', True):
+            expenses_attachments = attachments.filtered(lambda att: att.res_model == 'hr.expense')
+            if expenses_attachments:
+                expenses = self.env['hr.expense'].browse(expenses_attachments.mapped('res_id'))
+                for expense in expenses.filtered(lambda exp: bool(exp.sheet_id)):
+                    checksums = set(expense.sheet_id.attachment_ids.mapped('checksum'))
+                    for attachment in expense.attachment_ids.filtered(lambda att: att.checksum not in checksums):
+                        attachment.copy({
+                            'res_model': 'hr.expense.sheet',
+                            'res_id': expense.sheet_id.id,
+                        })
+        return attachments
+
+    def unlink(self):
+        res = True
+        if self.env.context.get('sync_attachment', True):
+            expenses_attachments = self.filtered(lambda att: att.res_model == 'hr.expense')
+            if expenses_attachments:
+                expenses = self.env['hr.expense'].browse(expenses_attachments.mapped('res_id'))
+                for expense in expenses.exists().filtered(lambda exp: bool(exp.sheet_id)):
+                    checksums = set(expense.attachment_ids.mapped('checksum'))
+                    attachments_to_unlink = expense.sheet_id.attachment_ids.filtered(lambda att: att.checksum in checksums)
```

Comment:
+1

Reply:
It's true the pattern is recurrent, but it's never actually the same.
Sometimes checksums are based on some overlap or difference of the attachments that needs to be considered (to avoid touching attachments owned by the sheet, double copies, or deleting attachments on other expense lines).
Same argument for the attachments we filter on, sometimes it's on those of the expense, other time on those of the sheet.

IMHO If I were to refactor this pattern in a method, it would provide little value, since both the attachments we filter on, and the checksums we consider for the filtering would need to be injected as params, and the method itself wouldn't provide any value besides hiding a filtered, and maybe giving a "name" to the pattern.

Unless you had something specific else in mind that I am missing?